crates/codegen/grammar/src/parser_definition.rs

```diff
@@ -53,6 +55,21 @@ impl Visitable for TriviaParserDefinitionRef {
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct DelimitedRecoveryOpts {
+    /// Whether completely unmatched body between the delimiters should
+    /// prevent the the error recovery from being applied.
+    pub disallow_unmatched_body: bool,
+}
+
+impl From<model::FieldDelimiters> for DelimitedRecoveryOpts {
+    fn from(delimiters: model::FieldDelimiters) -> Self {
+        Self {
+            disallow_unmatched_body: delimiters.disallow_unmatched_body.unwrap_or(false),
```

Comment:
I'm not sure I agree here. Using `std::Default` is the Rust-idiomatic way of communicating exactly this concept. And `bool` having a default value of `false` is understood and expected everywhere. So much that `unwrap_or_default()` and many other `*default*` methods have made it into the standard library.

I think wrapping the DSL type with another similar type that converts it into a similar `bool` value is anti-pattern here, and that using `FieldDelimiters::disallow_unmatched_body.unwrap_or_default()` directly is clear/understood by readers.

Reply:
We don't wrap the DSL type per se but rather we convert directly the domain configuration of `FieldDelimiters`, which has to have a specific shape because it's filled in by the user as expected by its schema (it forces a certain shape), into the (parser generator) domain-specific configuration option. Would consider this an anti-pattern?

Alternatively, how do you propose we pass this config down to the parser generator?

1. I didn't want to add an unnamed `bool` to the `ParserDefinitionNode::DelimitedBy` only to have to add context what the bool represents via inline comments.
2. I didn't want to change the enum variant shape from tuple-variant to struct-variant because it breaks the existing uniformity. Thinking about it, we probably should do that at some point, since it's not obvious now which members are braces and which is the body, but I'm leaving that out of this PR.
3. There's alternative of passing `FieldDelimiters` down into parser codegen but I'm not going to do that, since the other fields of `{open,close}_brace: Identifier` are redundant there.

Also, I don't have the energy to debate (nor want to), like we always do, the minutae of unimportant code like this and idiomaticness of Rust (do we really want to go there?).

What Rust does is return `false` in `Default for bool` because that's the least unexpected thing to do (zero value) and not because booleans are logically by default false in some general way, especially in the context of our logical default for this behaviour that we should explicitly specify here and shouldn't automatically be assumed to be false.

Additionally, the `unwrap_or_default` is a general mechanism for `Option<T> where T: Default` and `T` is arbitrary and can have custom impls; I fail to see why this being available generally means should be used everywhere else instead of `.unwrap_or(false)` here. There are multiple ways in the language or library to do the same thing and I'd expect to use the one that best communicates the intent and is easiest to read and understand.

This suggestion is also inconsistent with what we do elsewhere in the code, which is being explicit by default (heh) and often not relying on the defaults (i.e. recent PR with conventional shell completion file names debate, Clippy enforcing `Vec::default` instead of `Default::default` (`default_trait_access`) etc.).

I'm happy to change that to `unwrap_or_default` in this instance but it seems you had and still have a very specific implementation altogether in your mind. Do you want to take it over?

Reply:
> I fail to see why this being available generally means should be used everywhere else instead of .unwrap_or(false) here.

Because it means callsites (codegen crates) will have their own default values to set, which can change/diverge over time, introdocing bugs, rather than having a single source of truth (the default value).

This is my main concern, but it is not blocking the fix. Up to you if you feel strongly about it.

Reply:
Right now, the DSL only provides an option to specify a value by the user and it's still the parser generator that provides the default value, so we only have a single place that establishes the default.

I understand and can get behind this rationale, however in this instance if this is your main concern, then this further exemplifies my point that we should rely on a business logic that explicitly sets the defaults in a single place rather than relying on the fact that a `bool` value that will always have implicitly `false` as default.

Reply:
> we should rely on a business logic that explicitly sets the defaults in a single place

and that is `impl Default for T` (in that case `bool`) :) I think relying on the fact that it is `false` is fine, because it is a detail that callers of `unwrap_or_default()` never need to worry about. But to make the intention clearer, maybe we can use an enum in these cases, with `Yes` and `No` variants, and an explicit `#[default]` attribute? This would also make it possible to use in tuples without having to wrap as you mentioned above.

Reply:
I tried to account for both sides in https://github.com/NomicFoundation/slang/pull/862/commits/c09e7619b263a9e96976197de1ec1b0e2f3baeb4. We now have a derived `Default` for the options in a single place but we document a bit more why it's `false` and also it's used in the conversion code explicitly, so we don't risk diverging over time.

@OmarTawfik does this look good enough for you?

Reply:
I meant business logic as what Slang does, rather than what Rust's standard library does for all `bool`s. Surely we wouldn't use that default if we had a `recover_from_no_match` configuration option, that would need to default to true in our case?

> But to make the intention clearer...

Sounds good, however this is default is specifically tied to the delimiters; we can't leverage the existing `RecoverFromNoMatch` enum because:
1. we need it to default to false for semicolon recovery (without FIRST/FOLLOW it definitely is overeager)
2. for delimiters we want it to default it to true in 99% of the cases (except for this explicit opt-in, for now).

Do you think we'd have to introduce a separate `RecoverFromNoMatchInsideDelimiters` to encode that it's tied to the delimited recovery?

Reply:
Good point. In that case, what do you think about two different types, with two different defaults? Not only they would be documented differently (explaining these differences you mentioned), they would also be used uniformly throughout different codegen crates, getting a `Default` and acting on it, without having to understand the business logic behind it (encoded once in their respective `impl Default`). They would even be named differently, because terminator has a "prefix", not "contents".

Basically something like this:

```rust
pub struct FieldsErrorRecovery {
    pub terminated_by: Option<TerminatorErrorRecovery>,
    pub delimited_by: Option<DelimitersErrorRecovery>,
}

pub struct TerminatorErrorRecovery {
    pub terminator: Identifier,
    pub strategy: Option<TerminatorErrorRecoveryStrategy>,
}

#[derive(Default)]
pub enum TerminatorErrorRecoveryStrategy {
    SkipUnrecognizedPrefix,
    #[default]
    UnmatchParent,
}

pub struct DelimitersErrorRecovery {
    pub open: Identifier,
    pub close: Identifier,
    pub strategy: Option<DelimitersErrorRecoveryStrategy>,
}

#[derive(Default)]
pub enum DelimitersErrorRecoveryStrategy {
    #[default]
    SkipUnrecognizedContents,
    UnmatchParent,
}
```

But perhaps with less Java-ish names :)

Btw, my concern above are for the DSL types, and how they are used across the codebase. It is not as impactful for an internal parser type that is only used in one or two modules. Completely up to you if you feel strongly about either way.

Reply:
Good point. In that case, what do you think about two different types, with two different defaults? Not only they would be documented differently (explaining these differences you mentioned), they would also be used uniformly throughout different codegen crates, getting a `Default` and acting on it, without having to understand the business logic behind it (encoded once in their respective `impl Default`). They would even be named differently, because terminator has a "prefix", not "contents".

Basically something along the lines of:

```rust
pub struct FieldsErrorRecovery {
    pub terminated_by: Option<TerminatorErrorRecovery>,
    pub delimited_by: Option<DelimitersErrorRecovery>,
}

pub struct TerminatorErrorRecovery {
    pub terminator: Identifier,
    pub strategy: Option<TerminatorErrorRecoveryStrategy>,
}

#[derive(Default)]
pub enum TerminatorErrorRecoveryStrategy {
    SkipUnrecognizedPrefix,
    #[default]
    UnmatchParent,
}

pub struct DelimitersErrorRecovery {
    pub open: Identifier,
    pub close: Identifier,
    pub strategy: Option<DelimitersErrorRecoveryStrategy>,
}

#[derive(Default)]
pub enum DelimitersErrorRecoveryStrategy {
    #[default]
    SkipUnrecognizedContents,
    UnmatchParent,
}
```

Btw, my concern above are for the DSL types, and how they are used across the codebase. It is not as impactful for an internal parser type that is only used in one or two modules. Completely up to you if you feel strongly about either way.